File: designtime/webstudio/ui/src/app/editables/rule-template-instance/differ/ModificationEntry.ts
import { ModificationType } from './ModificationType';

import { RelatedFilterValueImpl } from '../../../editors/rule-template-instance-builder/RelatedFilterValueImpl';
import { SimpleFilterValueImpl } from '../../../editors/rule-template-instance-builder/SimpleFilterValueImpl';
import { SingleFilterImpl } from '../../../editors/rule-template-instance-builder/SingleFilterImpl';

export class ModificationEntry {
  private modificationType: ModificationType;
  private applied = false;
  private previousValue: string;
  private currentValue: string;
  private previousValObj: Object;
  private currentValObj: Object;

  constructor(modificationType: ModificationType,
    previousValue?: string,
    currentValue?: string,
    previousValObj?: Object,
    currentValObj?: Object) {
    this.modificationType = modificationType;
    if (previousValue) {
      this.previousValue = previousValue;
    }
    if (currentValue) {
      this.currentValue = currentValue;
    }
    if (previousValObj && currentValObj) {
      this.previousValObj = previousValObj;
      this.currentValObj = currentValObj;
      if (previousValObj instanceof SingleFilterImpl) {
        this.previousValue = ModificationEntry.getFilterAsString(<SingleFilterImpl>previousValObj);
      }
      if (currentValObj instanceof SingleFilterImpl) {
        this.currentValue = ModificationEntry.getFilterAsString(<SingleFilterImpl>currentValObj);
      }
    }
  }

  public static getFilterAsString(filter: SingleFilterImpl): string {
    if (filter == null) {
      return '';
    }
    let links = '';
    const relatedLinkImplList = filter.getLink();
    if (relatedLinkImplList) {
      for (let i = 0; i < relatedLinkImplList.length; i++) {
        links += (i === 0 ? ' ' : '\'s ') + relatedLinkImplList[i].getName();
      }
    }
    const operator: string = filter.getOperator();
    let filterValue = '';
    if (filter.getValue() instanceof SimpleFilterValueImpl) {
      const value: SimpleFilterValueImpl = <SimpleFilterValueImpl>filter.getValue();
      if (value) {
        filterValue = value.value;
      }
    } else if (filter.getValue() instanceof RelatedFilterValueImpl) {
      const value: RelatedFilterValueImpl = <RelatedFilterValueImpl>filter.getValue();
      const relatedLinkImplValueList = value.getLinks();
      for (let i = 0; i < relatedLinkImplValueList.length; i++) {
        filterValue += (i === 0 ? '' : '\'s ') + relatedLinkImplValueList[i].getName();
      }
    }
    return links + ' ' + operator + ' ' + filterValue;
  }

  public setModificationType(modificationType: ModificationType): void {
    this.modificationType = modificationType;
  }

  public getModificationType(): ModificationType {
    return this.modificationType;
  }

  public getPreviousValue(): string {
    return this.previousValue;
  }

  public setPreviousValue(previousValue: string): void {
    this.previousValue = previousValue;
  }

  public isApplied(): boolean {
    return this.applied;
  }

  public setApplied(applied: boolean): void {
    this.applied = applied;
  }

  public getCurrentValue(): string {
    return this.currentValue;
  }

  public setCurrentValue(currentValue: string): void {
    this.currentValue = currentValue;
  }

  public getPreviousValueObj(): Object {
    return this.previousValObj;
  }

  public getCurrentValueObj(): Object {
    return this.currentValObj;
  }
}

File: designtime/webstudio/WebStudio/src/com/tibco/cep/webstudio/client/diff/ModificationEntry.java
package com.tibco.cep.webstudio.client.diff;

import com.tibco.cep.webstudio.model.rule.instance.impl.SingleFilterImpl;

/**
 * 
 * @author moshaikh
 * 
 */
public class ModificationEntry {
	private ModificationType modificationType;
	private String previousValue;
	private boolean applied = false;
	private String currentValue;

	private Object previousValueObj;
	private Object currentValueObj;

	public ModificationEntry(ModificationType modificationType) {
		this.modificationType = modificationType;
	}

	public ModificationEntry(ModificationType modificationType,
			String previousValue) {
		this.modificationType = modificationType;
		this.previousValue = previousValue;
	}

	public ModificationEntry(ModificationType modificationType,
			String previousValue, String currentValue) {
		this(modificationType, previousValue);
		this.currentValue = currentValue;
	}

	public ModificationEntry(ModificationType modificationType, Object previousValueObj, Object currentValueObj) {
		this.previousValueObj = previousValueObj;
		this.currentValueObj = currentValueObj;
		if (previousValueObj instanceof SingleFilterImpl) {
			this.previousValue = DiffHelper.getFilterAsString((SingleFilterImpl) previousValueObj);
		}
		if (currentValueObj instanceof SingleFilterImpl) {
			this.currentValue = DiffHelper.getFilterAsString((SingleFilterImpl) currentValueObj);
		}
		this.modificationType = modificationType;
	}

	public void setModificationType(ModificationType modificationType) {
		this.modificationType = modificationType;
	}
	
	public ModificationType getModificationType() {
		return modificationType;
	}

	public String getPreviousValue() {
		return previousValue;
	}

	public void setPreviousValue(String previousValue) {
		this.previousValue = previousValue;
	}

	public boolean isApplied() {
		return applied;
	}

	public void setApplied(boolean applied) {
		this.applied = applied;
	}
	
	public String getCurrentValue() {
		return currentValue;
	}

	public void setCurrentValue(String currentValue) {
		this.currentValue = currentValue;
	}

	public Object getPreviousValueObj() {
		return previousValueObj;
	}

	public Object getCurrentValueObj() {
		return currentValueObj;
	}
}
